sql/tests: TestRandomSyntaxSQLSmith failed [expected FD to have candidate key]

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/12498400?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/12498400?buildTab=artifacts#/) on master @ [4d045594e8c65b56c82fcf2a1f14ee30cecfef3d](https://github.com/cockroachdb/cockroach/commits/4d045594e8c65b56c82fcf2a1f14ee30cecfef3d):

Random syntax error:

```
    rsg_test.go:911: Crash detected: server panic: pq: internal error: expected FD to have candidate key (1): key(1,2); (1)-->(2)
```
Query:

```
        WITH
        	"wi%f9th91275" (col550913, col550914)
        		AS (
        			SELECT
        				*
        			FROM
        				(
        					VALUES
        						((-26260):::INT8, 3440986062184580484:::INT8),
        						(30950:::INT8, 8601003195621751248:::INT8),
        						((-16662):::INT8, 6358180464743893851:::INT8)
        				)
        					AS tab224251 (col550913, col550914)
        			INTERSECT
        				SELECT
        					*
        				FROM
        					(
        						VALUES
        							(13478:::INT8, 8367685078544570941:::INT8),
        							(
        								(
        									WITH
        										"}wi😌th91274" ("Col550915")
        											AS (SELECT * FROM (VALUES (2241962435:::OID)) AS "tAb224252" ("Col550915"))
        									SELECT
        										t̔ab224255._int4 AS col550917
        									FROM
        										defaultdb.public.seed@seed__int8__float8__date_idx AS tab224253
        										JOIN defaultdb.public."😔ta ble1"@[0] AS "}tab224254"
        											JOIN defaultdb.public.seed@seed__int8__float8__date_idx AS t̔ab224255 ON
        													("}tab224254".crdb_internal_mvcc_timestamp)
        													= (t̔ab224255.crdb_internal_mvcc_timestamp) ON
        												(tab224253.tableoid) = ("}tab224254"."co
l1_9")
        												AND (tab224253._int8) = (t̔ab224255._int4)
        									WHERE
        										EXISTS(
        											SELECT
        												NULL AS col550916
        											FROM
        												defaultdb.public."ta\\xf7ble2"@"ta\\xf7ble2_expr_col2_\\u834311_col2_10_c'ol2_13_expr1_col2_12_expr2_ col2_9_col2_
8_col2_5_%col2_7_col2_%p3_col2_}14_co{l2_4_idx"
        													AS "t'ab224256",
        												defaultdb.public."ta\\xf7ble2"@[0] AS tab224257,
        												defaultdb.public.seed@seed__int8__float8__date_idx AS tab224258
        												RIGHT JOIN defaultdb.public.seed@[0] AS "
        tab224259" ON
        														("
        tab224259"._bool AND true),
        												defaultdb.public.seed@seed__int8__float8__date_idx AS t̨ab224260
        											LIMIT
        												88:::INT8
        										)
        									LIMIT
        										1:::INT8
        								),
        								((-7248027343665288937):::INT8::INT8 << (-6792667406989627650):::INT8::INT8)::INT8
        							),
        							(NULL, (-4871576753501471461):::INT8),
        							((-21668):::INT8, (-2842918276145933880):::INT8),
        							(NULL, 6758530915389657721:::INT8),
        							((-1217):::INT8, 755602209118446107:::INT8)
        					)
        						AS " t😟ab%v224261" ("col\\x3a550918", "c\\xf4ol550919")
        			INTERSECT ALL
        				SELECT
        					"t ab224265"._int8 AS "\rcol550920", " tab224266"."col
        1_18" AS "col 550921"
        				FROM
        					defaultdb.public."ta\\xf7ble2"@"ta\\xf7ble2_expr_col2_\\u834311_col2_10_c'ol2_13_expr1_col2_12_expr2_ col2_9_col2_
8_col2_5_%col2_7_col2_%p3_col2_}14_co{l2_4_idx"
        						AS "t\\U00071ACCab224262",
        					defaultdb.public."😔ta ble1" AS tab224263
        					JOIN defaultdb.public."😔ta ble1" AS tab224264 ON
        							(tab224263.col1_1) = (tab224264.col1_1) AND (tab224263."col'1_12") = (tab224264."col'1_12"),
        					defaultdb.public.seed@seed__int8__float8__date_idx AS "t ab224265"
        					JOIN defaultdb.public."😔ta ble1"@[0] AS " tab224266" ON
        							("t ab224265"._int8) = (" tab224266"."c ol1_15")
        				WHERE
        					"t ab224265"._bool
        		),
        	"with!91276" (col550922, col550923)
        		AS (SELECT * FROM (VALUES (NULL, 0.4761936068534851:::FLOAT8)) AS "t?%vab224267" (col550922, col550923)),
        	with91277 ("CͅoL550924")
        		AS (
        			SELECT
        				*
        			FROM
        				(VALUES (585443625:::OID), (492361766:::OID), (2536711631:::OID), (1939953041:::OID))
        					AS "%vtab224268" ("CͅoL550924")
        		)
        SELECT
        	tab224269.tableoid AS col550925
        FROM
        	defaultdb.public."😔ta ble1"@"😔ta ble1_expr_key" AS tab224269
        GROUP BY
        	tab224269.tableoid
        ORDER BY
        	tab224269.tableoid ASC NULLS LAST
        LIMIT
        	10:::INT8;
```
Schema:

```
    rsg_test.go:720: To reproduce, use schema:
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;;
    rsg_test.go:722: CREATE TABLE "😔ta ble1" (col1_0 TIME NOT NULL, col1_1 OID NULL, co͎l1_2 REGTYPE, col1_3 REGTYPE NULL, "col(1
        😬_4" TSVECTOR NOT NULL, col😙1_5 FLOAT4 NOT NULL, col1😍_6 REGROLE NULL, col1_7 GEOGRAPHY NOT NULL, col1_8 BOX2D NOT NULL, "co
l1_9" REGROLE, col1_10 INET NOT NULL, col1😜_11 TIMESTAMP, "col'1_12" UUID NOT NULL, col1_13 OID, col1_14 STRING NULL, "c ol1_15" INT8 NOT NULL, "col1'_16" REGTYPE NULL, "\\u0822c�ol1 _17" INET NOT NULL, "col
        1_18" INT8 NOT NULL AS ("c ol1_15" + 9223372036854775807:::INT8) VIRTUAL, PRIMARY KEY ("col'1_12" ASC), UNIQUE (lower(CAST(col1_0 AS STRING)) DESC) WHERE ((((("😔ta ble1".col1_14 > '':::STRING) OR ("😔ta ble1"."col
        1_18" > 9223372036854775807:::INT8)) OR ("😔ta ble1".col1😜_11 > '0001-01-01 00:00:00':::TIMESTAMP)) OR ("😔ta ble1".col😙1_5 < 0.0:::FLOAT8)) OR ("😔ta ble1".col1_0 > '24:00:00':::TIME)) OR ("😔ta ble1"."c ol1_15" <= (-2147483648):::INT8), INDEX (col1_1 ASC) STORING (col1_0, co͎l1_2, col1_3, "col(1
        😬_4", col1_13, col1_14, "c ol1_15", "col1'_16", "\\u0822c�ol1 _17"), INDEX (col😙1_5 DESC), INDEX ((CASE WHEN col1😜_11 IS NULL THEN e'xOn\x7f\x13':::STRING ELSE e'_\x1f?Fkq-':::STRING END), "co
l1_9") STORING (col1_0, "col(1
        😬_4", col😙1_5, col1_7, "\\u0822c�ol1 _17"), INVERTED INDEX (col1_14 gin_trgm_ops ASC), INDEX ("c ol1_15" DESC, col1😍_6 ASC) STORING (col1_1, col1_7, "co
l1_9", col1😜_11, col1_14, "col1'_16") WHERE ((((("😔ta ble1".col1_0 <= '24:00:00':::TIME) OR ("😔ta ble1".col1😜_11 > '3000-01-01 00:00:00':::TIMESTAMP)) AND ("😔ta ble1".col😙1_5 < '+Inf':::FLOAT8)) OR ("😔ta ble1".col1_14 >= '':::STRING)) OR ("😔ta ble1"."c ol1_15" >= 2147483647:::INT8)) AND ("😔ta ble1"."col
        1_18" < 1:::INT8), UNIQUE ("\\u0822c�ol1 _17", "c ol1_15" DESC, "col'1_12" ASC) WHERE ((("😔ta ble1"."c ol1_15" < 1:::INT8) OR ("😔ta ble1".col1_0 = '24:00:00':::TIME)) OR ("😔ta ble1".col1_14 = e'\U00002603':::STRING)) OR ("😔ta ble1".col😙1_5 != 0.0:::FLOAT8), INDEX (col1😜_11 DESC, (CASE WHEN col1😍_6 IS NULL THEN 'X0':::STRING ELSE e'S,G\x01]':::STRING END)) WHERE ((((("😔ta ble1".col1😜_11 >= '4714-11-24 00:00:00 BC':::TIMESTAMP) AND ("😔ta ble1"."c ol1_15" = 2147483647:::INT8)) AND ("😔ta ble1"."col
        1_18" <= (-2147483648):::INT8)) AND ("😔ta ble1".col1_14 > 'X':::STRING)) AND ("😔ta ble1".col1_0 = '24:00:00':::TIME)) AND ("😔ta ble1".col😙1_5 >= 0.0:::FLOAT8) NOT VISIBLE, FAMILY (col1_0), FAMILY (col1_1), FAMILY (col1_10, "col1'_16", col1😜_11), FAMILY ("\\u0822c�ol1 _17", col1_13, col1_14), FAMILY (col😙1_5, "co
l1_9", "c ol1_15"), FAMILY (col1_3, co͎l1_2), FAMILY (col1_8, "col'1_12", col1😍_6, "col(1
        😬_4", col1_7));
    rsg_test.go:722: CREATE TABLE "ta\\xf7ble2" (col2_0 REFCURSOR NOT NULL, col2_1 NAME NOT NULL, "co l2_2" GEOMETRY NOT NULL, "col2_%p3" PG_LSN NOT NULL, "co{l2_4" BYTES NOT NULL, col2_5 BOX2D NOT NULL, c̣ol2_6 FLOAT8 NOT NULL, "%col2_7" INTERVAL[] NOT NULL, "col2_
8" STRING NOT NULL AS (lower(col2_1)) VIRTUAL, " col2_9" STRING NOT NULL AS (lower(CAST("co l2_2" AS STRING))) STORED, col2_10 FLOAT8 NOT NULL AS (c̣ol2_6 + (-1.0810193635320235):::FLOAT8) VIRTUAL, "col2_\\u834311" STRING NOT NULL AS (lower(CAST("col2_%p3" AS STRING))) STORED, col2_12 STRING NOT NULL AS (lower(CAST("col2_%p3" AS STRING))) STORED, "c'ol2_13" STRING NOT NULL AS (lower(CAST("co l2_2" AS STRING))) VIRTUAL, "col2_}14" STRING NOT NULL AS (CASE WHEN "co{l2_4" IS NULL THEN e'l\x1c':::STRING ELSE e'E\x075S0':::STRING END) VIRTUAL, INDEX (lower(CAST(col2_0 AS STRING)), "col2_\\u834311" DESC, col2_10 ASC, "c'ol2_13" DESC, lower(CAST("co l2_2" AS STRING)) ASC, col2_12 ASC, lower(col2_1) ASC, " col2_9" DESC, "col2_
8" DESC, col2_5 ASC, "%col2_7", "col2_%p3", "col2_}14" DESC, "co{l2_4") WHERE ((("ta\\xf7ble2"."co{l2_4" >= '\x00':::BYTES) AND ("ta\\xf7ble2".col2_12 >= '"':::STRING)) AND ("ta\\xf7ble2"."col2_}14" >= '':::STRING)) OR ("ta\\xf7ble2"."c'ol2_13" >= e'\x00':::STRING), UNIQUE ("c'ol2_13", "co{l2_4" DESC) STORING (col2_0, "co l2_2", col2_5, "col2_\\u834311", col2_12), FAMILY ("col2_\\u834311"), FAMILY ("%col2_7", col2_12, "col2_%p3"), FAMILY (col2_5), FAMILY (c̣ol2_6), FAMILY (" col2_9", "co{l2_4"), FAMILY (col2_0, "co l2_2", col2_1));
    rsg_test.go:722: ALTER TABLE "😔ta ble1" INJECT STATISTICS e'[{"avg_size": 4, "columns": ["col1_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 342182347033092440, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 12, "columns": ["col1_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4840346099999827605, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 31, "columns": ["\\"col\'1_12\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6361700926801116724, "histo_buckets": [{"distinct_range": 0, "num_eq": 5000, "num_range": 0, "upper_bound": "2deb76c7-80d6-4db3-afee-49e724895d64"}, {"distinct_range": 4443243875766344700, "num_eq": 2000, "num_range": 7564325868882269892, "upper_bound": "5733d0a8-c110-4f52-881b-d7084965a760"}, {"distinct_range": 3538440516453493000, "num_eq": 8681619698994075112, "num_range": 3538440516453493006, "upper_bound": "8aca1442-185a-4c3b-b245-f99bb907357f"}, {"distinct_range": 3000000000, "num_eq": 0, "num_range": 3000000000, "upper_bound": "bece748f-c6c1-4716-a499-762f178230e1"}, {"distinct_range": 0, "num_eq": 4000000000, "num_range": 0, "upper_bound": "f737b5ce-b8a3-4337-a18a-240b3cdee7b6"}, {"distinct_range": 0, "num_eq": 7438363271843399279, "num_range": 0, "upper_bound": "ffb30230-009b-4faa-a136-2d34472982b9"}], "histo_col_type": "UUID", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 18, "columns": ["col1_13"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2394790280193904548, "histo_col_type": "", "name": "__auto__", "null_count": 2669226602733356485, "row_count": 7111607836003650476}, {"avg_size": 18, "columns": ["\\"c ol1_15\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6831837239125100830, "histo_buckets": [{"distinct_range": 0, "num_eq": 3092155316031585651, "num_range": 0, "upper_bound": "-4264729792013734721"}, {"distinct_range": 0, "num_eq": 8000, "num_range": 787467885685289096, "upper_bound": "-2147483648"}, {"distinct_range": 1761886.3036653523, "num_eq": 40000, "num_range": 3000000, "upper_bound": "2837890789512970907"}, {"distinct_range": 0, "num_eq": 4000, "num_range": 0, "upper_bound": "4986507291715339396"}, {"distinct_range": 0.41804348587198054, "num_eq": 30000000000, "num_range": 40, "upper_bound": "6705553158749660653"}], "histo_col_type": "INT8", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 6, "columns": ["\\"\\\\\\\\u0822c\\u000bol1 _17\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6718507378004846743, "histo_buckets": [{"distinct_range": 0, "num_eq": 7158117771053396521, "num_range": 0, "upper_bound": "243.240.105.237/26"}, {"distinct_range": 3690482317262352400, "num_eq": 2000000000, "num_range": 4537295378446735415, "upper_bound": "1bcc:3b36:fa28:e274:bc25:a143:9350:c700/21"}, {"distinct_range": 0, "num_eq": 6081323262820347814, "num_range": 1915360774432470622, "upper_bound": "6eef:4f87:a774:2d1:efe0:bc7c:3b96:df58/98"}, {"distinct_range": 0, "num_eq": 10000000000, "num_range": 2513947901700233293, "upper_bound": "dac:4844:79d9:5ba6:b901:a491:e3e5:6840/114"}], "histo_col_type": "INET", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 7, "columns": ["col\U0001F6191_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 655495596891575576, "histo_buckets": [{"distinct_range": 0, "num_eq": 1010814541012364937, "num_range": 0, "upper_bound": "-1.1875314712524414"}, {"distinct_range": 575165890164545700, "num_eq": 100, "num_range": 2950614113666503276, "upper_bound": "-0.8376743197441101"}, {"distinct_range": 0, "num_eq": 5128815377950459150, "num_range": 7121641997941810727, "upper_bound": "-0.31200870871543884"}, {"distinct_range": 7162262739343573000, "num_eq": 4322903422939441098, "num_range": 7162262739343573012, "upper_bound": "0.18045736849308014"}, {"distinct_range": 0, "num_eq": 1558730753734658767, "num_range": 1708332510671412633, "upper_bound": "0.8379345536231995"}], "histo_col_type": "FLOAT4", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 21, "columns": ["\\"col1\'_16\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4197370423036294259, "histo_col_type": "", "name": "__auto__", "null_count": 1423542703653952126, "row_count": 7111607836003650476}, {"avg_size": 29, "columns": ["col1_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2312928793003825655, "histo_col_type": "", "name": "__auto__", "null_count": 2035220349278981773, "row_count": 7111607836003650476}, {"avg_size": 7, "columns": ["\\"col(1\\n\U0001F62C_4\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6506804490450153520, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 2, "columns": ["col1\U0001F60D_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1541937641869476319, "histo_col_type": "", "name": "__auto__", "null_count": 4555932475859806721, "row_count": 7111607836003650476}, {"avg_size": 13, "columns": ["col1_8"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5553621303427411458, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 24, "columns": ["\\"col\\n1_18\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5248371000675814893, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 26, "columns": ["col1_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 248021091428556465, "histo_buckets": [{"distinct_range": 0, "num_eq": 0, "num_range": 0, "upper_bound": "0"}, {"distinct_range": 0, "num_eq": 6714624831033978592, "num_range": 4000000000, "upper_bound": "3073110212"}, {"distinct_range": 80000000, "num_eq": 500, "num_range": 80000000, "upper_bound": "3685793425"}, {"distinct_range": 3000, "num_eq": 200, "num_range": 3000, "upper_bound": "3696286551"}, {"distinct_range": 343360619112810560, "num_eq": 5526532564975432525, "num_range": 1267322895489772892, "upper_bound": "3944742510"}], "histo_col_type": "OID", "name": "__auto__", "null_count": 3884372638703307966, "row_count": 7111607836003650476}, {"avg_size": 2, "columns": ["co\u034El1_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6120841782925906843, "histo_col_type": "", "name": "__auto__", "null_count": 1966171257385565931, "row_count": 7111607836003650476}, {"avg_size": 6, "columns": ["\\"co\\rl1_9\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1142823950936462338, "histo_col_type": "", "name": "__auto__", "null_count": 4650685169659412899, "row_count": 7111607836003650476}, {"avg_size": 22, "columns": ["col1_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4138189113918657386, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 7111607836003650476}, {"avg_size": 8, "columns": ["col1\U0001F61C_11"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2708074816199592990, "histo_buckets": [{"distinct_range": 0, "num_eq": 500000000, "num_range": 0, "upper_bound": "1999-07-11 08:30:50.000875"}, {"distinct_range": 2000000, "num_eq": 1000000000, "num_range": 2000000, "upper_bound": "2018-02-14 17:46:36.000899"}, {"distinct_range": 9.4489213157423, "num_eq": 5075696827330074992, "num_range": 20, "upper_bound": "2018-06-18 17:23:01.000542"}, {"distinct_range": 70000, "num_eq": 100, "num_range": 70000, "upper_bound": "2022-12-03 16:39:08.00055"}, {"distinct_range": 0, "num_eq": 5358349823791646377, "num_range": 7000000, "upper_bound": "2024-03-19 20:11:32.000436"}, {"distinct_range": 50000000, "num_eq": 8103698289577193064, "num_range": 50000000, "upper_bound": "2026-11-15 23:15:13.000292"}, {"distinct_range": 45412313069.00661, "num_eq": 7638900820152748423, "num_range": 50000000000, "upper_bound": "2032-02-20 21:39:39.000588"}], "histo_col_type": "TIMESTAMP", "name": "__auto__", "null_count": 1313639758171246969, "row_count": 7111607836003650476}, {"avg_size": 6, "columns": ["col1_14"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3675030585564694057, "histo_col_type": "", "name": "__auto__", "null_count": 5167317451403802610, "row_count": 7111607836003650476}]':::JSONB;
    rsg_test.go:722: ALTER TABLE "ta\\xf7ble2" INJECT STATISTICS e'[{"avg_size": 16, "columns": ["col2_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 6802777703, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 21, "columns": ["col2_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2518395221, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 13, "columns": ["\\"col2_\\r8\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 17324004, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 14, "columns": ["\\" col2_9\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 8804439540, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 0, "columns": ["col2_12"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2385136442, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 0, "columns": ["\\"c\'ol2_13\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 9421445014, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 11, "columns": ["col2_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5399925661, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 27, "columns": ["\\"co l2_2\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7144209403, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 20, "columns": ["\\"col2_\\\\\\\\u834311\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5184637274, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 28, "columns": ["\\"col2_%p3\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 735013102, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 17, "columns": ["\\"co{l2_4\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2525351515, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 30, "columns": ["c\u0323ol2_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 651541511, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 29, "columns": ["\\"%col2_7\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7527245266, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 8, "columns": ["col2_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4487678487, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}, {"avg_size": 9, "columns": ["\\"col2_}14\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 8521066706, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 10000000000}]':::JSONB;
    rsg_test.go:722: CREATE TYPE rand_typ_0 AS ENUM ('clb', 'shcpjx', 'sko', 'ml', 'c', 'hwz');
    rsg_test.go:722: CREATE TYPE rand_typ_1 AS ENUM ('vyufzd');
    rsg_test.go:722: CREATE TYPE rand_typ_2 AS ENUM ('zjjo', 'jic', 'oldsmg');
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: 
        BEGIN; CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning'); COMMIT;
        BEGIN;
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        COMMIT;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (7046321948781838705);
    rsg_test.go:722: ALTER TABLE seed SCATTER;
    rsg_test.go:724: 
    rsg_test.go:725: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxSQLSmith169170792
--- FAIL: TestRandomSyntaxSQLSmith (380.79s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33176

--- a/pkg/sql/opt/memo/testdata/logprops/set
+++ b/pkg/sql/opt/memo/testdata/logprops/set
@@ -10,6 +10,10 @@ exec-ddl
 CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT)
 ----
 
+exec-ddl
+CREATE TABLE mn (m INT, n INT, PRIMARY KEY (m, n))
+----
+
 build
 SELECT * FROM xy UNION SELECT * FROM uv
 ----
@@ -639,3 +643,29 @@ except-all
       ├── prune: (2)
       └── tuple [type=tuple{int}]
            └── const: 1 [type=int]
+
+# Regression test for #113817. The key from the LHS of the intersection should
+# be reduced after adding the FDs from the RHS.
+norm
+SELECT m, n FROM mn
+INTERSECT
+SELECT x, y FROM xy
+----
+intersect-all
+ ├── columns: m:1(int!null) n:2(int)
+ ├── left columns: m:1(int!null) n:2(int)
+ ├── right columns: x:5(int) y:6(int)
+ ├── key: (1)
+ ├── fd: (1)-->(2)
+ ├── interesting orderings: (+1,+2)
+ ├── scan mn
+ │    ├── columns: m:1(int!null) n:2(int!null)
+ │    ├── key: (1,2)
+ │    ├── prune: (1,2)
+ │    └── interesting orderings: (+1,+2)
+ └── scan xy
+      ├── columns: x:5(int!null) y:6(int)
+      ├── key: (5)
+      ├── fd: (5)-->(6)
+      ├── prune: (5,6)
+      └── interesting orderings: (+5)

--- a/pkg/sql/opt/props/func_dep.go
+++ b/pkg/sql/opt/props/func_dep.go
@@ -1152,6 +1152,7 @@ func (f *FuncDepSet) AddFrom(fdset *FuncDepSet) {
 		fd := &fdset.deps[i]
 		f.addDependency(fd.from, fd.to, fd.strict, fd.equiv)
 	}
+	f.tryToReduceKey(opt.ColSet{} /* notNullCols */)
 }
 
 // AddEquivFrom is similar to AddFrom, except that it only adds equivalence

--- a/pkg/sql/opt/props/func_dep_test.go
+++ b/pkg/sql/opt/props/func_dep_test.go
@@ -795,6 +795,17 @@ func TestFuncDeps_AddFrom(t *testing.T) {
 	abcde.AddStrictKey(c(2, 3), c(1, 2, 3, 4, 5))
 	verifyFD(t, abcde, "key(2,3); (1)-->(2-5), (2,3)-->(1,4,5)")
 	testColsAreStrictKey(t, abcde, c(1), true)
+
+	// Add a single-column strict dependency to a multi-column strict
+	// dependency.
+	m := &props.FuncDepSet{}
+	m.AddStrictKey(c(1, 2), c(1, 2))
+	verifyFD(t, m, "key(1,2)")
+	s := &props.FuncDepSet{}
+	s.AddStrictKey(c(1), c(1, 2))
+	verifyFD(t, s, "key(1); (1)-->(2)")
+	m.AddFrom(s)
+	verifyFD(t, m, "key(1); (1)-->(2)")
 }
 
 func TestFuncDeps_AddEquivFrom(t *testing.T) {
@@ -815,6 +826,16 @@ func TestFuncDeps_AddEquivFrom(t *testing.T) {
 	equiv.ProjectCols(opt.ColSet{})
 	equiv.AddEquivFrom(product)
 	verifyFD(t, &equiv, "(1)==(10), (10)==(1), (2)==(11), (11)==(2)")
+
+	// Add an equivalency to a multi-column strict dependency.
+	m := &props.FuncDepSet{}
+	m.AddStrictKey(c(1, 2), c(1, 2))
+	verifyFD(t, m, "key(1,2)")
+	s := &props.FuncDepSet{}
+	s.AddEquivalency(1, 2)
+	verifyFD(t, s, "(1)==(2), (2)==(1)")
+	m.AddEquivFrom(s)
+	verifyFD(t, m, "key(2); (1)==(2), (2)==(1)")
 }
 
 func TestFuncDeps_MakeProduct(t *testing.T) {

